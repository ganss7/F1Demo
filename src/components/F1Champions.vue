<template>
  <div class="champions-list">
    <md-content class="heading">
      <h1>F1 Champions from {{yearFrom}} to {{yearTo}}:</h1>  
    </md-content>
    
    <md-progress-bar v-if="!championsList.length" md-mode="query"></md-progress-bar>
    <md-list>
      <transition-group
        @before-enter="beforeEnter"
        @enter="enter"
        @after-enter="afterEnter"
        @before-leave="beforeLeave"
        @leave="leave"
        @afterLeave="afterLeave"
        tag="div">
        <md-list-item v-for="(champion, index) in championsList" :key="champion.season"
          :to="{ name: 'season', params: { season: champion.season, winner: champion.DriverStandings[0].Driver } }"
          :data-index="index">
          {{index+1}}.&nbsp;<strong>{{champion.season}}</strong>&nbsp;
          -&nbsp;
          <strong>{{getDriverFullName(champion.DriverStandings[0].Driver)}}</strong>
        </md-list-item>
      </transition-group>
    </md-list>

    <!-- dialog for showing error messages -->
    <md-dialog :md-active.sync="showErrorMessage">
      <md-dialog-title>Error</md-dialog-title>
      <md-dialog-content>{{errorMessage}}</md-dialog-content>
      <md-dialog-actions>
        <md-button @click="showErrorMessage = false">Ok</md-button>
      </md-dialog-actions>
    </md-dialog>
  </div>
</template>

<script>
import f1history from '@/api/f1history'
import utils from '@/mixins/utils'
import listAnimation from '@/mixins/listAnimation'

export default {
  props: {
    yearFrom: String,
    yearTo: String
  },
  mixins: [utils, listAnimation],
  data () {
    return {
      championsList: [],
      showErrorMessage: false,
      errorMessage: ''
    }
  },
  created () {
    f1history.getChampionsList({yearFrom: this.yearFrom, yearTo: this.yearTo})
      .then(data => {
        this.championsList = data.MRData.StandingsTable.StandingsLists
        this.$options.listLength = this.championsList.length
      })
      .catch(error => {
        this.errorMessage = error.message
        this.showErrorMessage = true
      }) 
  }
}
</script>


