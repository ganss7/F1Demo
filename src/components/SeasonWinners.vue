<template>
  <div class="season-winners-list">
    <md-content class="heading">
      <h1>Season {{season}} race winners:</h1>
    </md-content>

    <md-progress-bar v-if="!winnersList.length" md-mode="query"></md-progress-bar>
    <md-list>
      <transition-group
        @before-enter="beforeEnter"
        @enter="enter"
        @after-enter="afterEnter"
        @before-leave="beforeLeave"
        @leave="leave"
        @afterLeave="afterLeave"
        tag="div">
        <md-list-item v-for="(winner, index) in winnersList" :key="winner.round"
          :class="`${myWinner &&  myWinner.driverId === winner.Results[0].Driver.driverId ? 'highlight' : ''}`"
          :data-index="index">
          {{index+1}}.&nbsp;<strong>{{winner.raceName}}</strong>
          &nbsp;-&nbsp;
          <a :href="winner.Results[0].Driver.url" target="_blank">{{getDriverShortName(winner.Results[0].Driver)}}</a>
        </md-list-item>
      </transition-group>
    </md-list>

    <!-- dialog for showing error messages -->
    <md-dialog :md-active.sync="showErrorMessage">
      <md-dialog-title>Error</md-dialog-title>
      <md-dialog-content>{{errorMessage}}</md-dialog-content>
      <md-dialog-actions>
        <md-button @click="showErrorMessage = false">Ok</md-button>
      </md-dialog-actions>
    </md-dialog>
  </div>
</template>

<script>
import f1history from '@/api/f1history'
import utils from '@/mixins/utils'
import listAnimation from '@/mixins/listAnimation'

export default {
  props: {
    season: String,
    winner: Object
  },
  mixins: [utils, listAnimation],
  data () {
    return {
      winnersList: [],
      myWinner: this.winner,
      showErrorMessage: false,
      errorMessage: ''
    }
  },
  created () {
    f1history.getSeasonWinnersList(this.season)
      .then(data => this.winnersList = data.MRData.RaceTable.Races)
      .catch(error => {
        this.errorMessage = error.message
        this.showErrorMessage = true
      }) 
    
    if (!this.myWinner) {
      f1history.getSeasonWinner(this.season)
        .then(data => this.myWinner = data.MRData.StandingsTable.StandingsLists[0].DriverStandings[0].Driver)
        .catch(error => {
          this.errorMessage = error.message
          this.showErrorMessage = true
        })
    }
  }
}
</script>

<style lang="scss" scoped>
  #app .md-list-item.highlight {
    background: var(--md-theme-default-accent);
  }
</style>

