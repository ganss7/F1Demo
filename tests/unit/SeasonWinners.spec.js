import { shallowMount, createLocalVue  } from '@vue/test-utils'

import { MdApp, MdToolbar, MdContent, MdList, MdDialog, MdButton, MdProgress } from 'vue-material/dist/components'
import SeasonWinners from '@/components/SeasonWinners.vue'

const localVue = createLocalVue()

localVue.use(MdApp)
localVue.use(MdToolbar)
localVue.use(MdContent)
localVue.use(MdList)
localVue.use(MdDialog)
localVue.use(MdButton)
localVue.use(MdProgress)

jest.mock('@/api/f1history')

const inputProps = {
  season: '2005',
  winner: {"driverId":"alonso","permanentNumber":"14","code":"ALO","url":"http:\/\/en.wikipedia.org\/wiki\/Fernando_Alonso","givenName":"Fernando","familyName":"Alonso","dateOfBirth":"1981-07-29","nationality":"Spanish"}
}

describe('SeasonWinners.vue', () => {

  it('highlights races won by champion', (done) => {

    const wrapper = shallowMount(SeasonWinners, {
      localVue,
      propsData: inputProps
    })

    wrapper.vm.$nextTick(() => {
      const racesList = wrapper.findAll('mdlistitem-stub')
      const wonByChampion = racesList.filter(x => x.text().includes(wrapper.vm.winner.familyName))
      const highlighted = racesList.filter(x => x.is('.highlight') )

      expect(wonByChampion).toEqual(highlighted)
      done()
    })

  })

  it('renders correctrly', (done) => {

    const wrapper = shallowMount(SeasonWinners, {
      localVue,
      propsData: inputProps
    })

    wrapper.vm.$nextTick(() => {
      expect(wrapper.element).toMatchSnapshot()
      done()
    })

  })

})
