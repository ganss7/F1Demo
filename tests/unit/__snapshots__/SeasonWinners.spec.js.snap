// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeasonWinners.vue renders correctrly 1`] = `
<div
  class="season-winners-list"
>
  <mdcontent-stub
    class="heading"
    mdtag="div"
  >
    <h1>
      Season 2005 race winners:
    </h1>
  </mdcontent-stub>
   
  <!---->
   
  <mdlist-stub>
    <span
      tag="div"
    >
      <mdlistitem-stub
        class=""
        data-index="0"
        dataindex="0"
      >
        
        1. 
        <strong>
          Australian Grand Prix
        </strong>
        
         - 
        
        <a
          href="http://en.wikipedia.org/wiki/Giancarlo_Fisichella"
          target="_blank"
        >
          G. Fisichella
        </a>
      </mdlistitem-stub>
      <mdlistitem-stub
        class="highlight"
        data-index="1"
        dataindex="1"
      >
        
        2. 
        <strong>
          Malaysian Grand Prix
        </strong>
        
         - 
        
        <a
          href="http://en.wikipedia.org/wiki/Fernando_Alonso"
          target="_blank"
        >
          F. Alonso
        </a>
      </mdlistitem-stub>
      <mdlistitem-stub
        class="highlight"
        data-index="2"
        dataindex="2"
      >
        
        3. 
        <strong>
          Bahrain Grand Prix
        </strong>
        
         - 
        
        <a
          href="http://en.wikipedia.org/wiki/Fernando_Alonso"
          target="_blank"
        >
          F. Alonso
        </a>
      </mdlistitem-stub>
      <mdlistitem-stub
        class="highlight"
        data-index="3"
        dataindex="3"
      >
        
        4. 
        <strong>
          San Marino Grand Prix
        </strong>
        
         - 
        
        <a
          href="http://en.wikipedia.org/wiki/Fernando_Alonso"
          target="_blank"
        >
          F. Alonso
        </a>
      </mdlistitem-stub>
      <mdlistitem-stub
        class=""
        data-index="4"
        dataindex="4"
      >
        
        5. 
        <strong>
          Spanish Grand Prix
        </strong>
        
         - 
        
        <a
          href="http://en.wikipedia.org/wiki/Kimi_R%C3%A4ikk%C3%B6nen"
          target="_blank"
        >
          K. Räikkönen
        </a>
      </mdlistitem-stub>
      <mdlistitem-stub
        class=""
        data-index="5"
        dataindex="5"
      >
        
        6. 
        <strong>
          Monaco Grand Prix
        </strong>
        
         - 
        
        <a
          href="http://en.wikipedia.org/wiki/Kimi_R%C3%A4ikk%C3%B6nen"
          target="_blank"
        >
          K. Räikkönen
        </a>
      </mdlistitem-stub>
      <mdlistitem-stub
        class="highlight"
        data-index="6"
        dataindex="6"
      >
        
        7. 
        <strong>
          European Grand Prix
        </strong>
        
         - 
        
        <a
          href="http://en.wikipedia.org/wiki/Fernando_Alonso"
          target="_blank"
        >
          F. Alonso
        </a>
      </mdlistitem-stub>
      <mdlistitem-stub
        class=""
        data-index="7"
        dataindex="7"
      >
        
        8. 
        <strong>
          Canadian Grand Prix
        </strong>
        
         - 
        
        <a
          href="http://en.wikipedia.org/wiki/Kimi_R%C3%A4ikk%C3%B6nen"
          target="_blank"
        >
          K. Räikkönen
        </a>
      </mdlistitem-stub>
      <mdlistitem-stub
        class=""
        data-index="8"
        dataindex="8"
      >
        
        9. 
        <strong>
          United States Grand Prix
        </strong>
        
         - 
        
        <a
          href="http://en.wikipedia.org/wiki/Michael_Schumacher"
          target="_blank"
        >
          M. Schumacher
        </a>
      </mdlistitem-stub>
      <mdlistitem-stub
        class="highlight"
        data-index="9"
        dataindex="9"
      >
        
        10. 
        <strong>
          French Grand Prix
        </strong>
        
         - 
        
        <a
          href="http://en.wikipedia.org/wiki/Fernando_Alonso"
          target="_blank"
        >
          F. Alonso
        </a>
      </mdlistitem-stub>
      <mdlistitem-stub
        class=""
        data-index="10"
        dataindex="10"
      >
        
        11. 
        <strong>
          British Grand Prix
        </strong>
        
         - 
        
        <a
          href="http://en.wikipedia.org/wiki/Juan_Pablo_Montoya"
          target="_blank"
        >
          J. Pablo Montoya
        </a>
      </mdlistitem-stub>
      <mdlistitem-stub
        class="highlight"
        data-index="11"
        dataindex="11"
      >
        
        12. 
        <strong>
          German Grand Prix
        </strong>
        
         - 
        
        <a
          href="http://en.wikipedia.org/wiki/Fernando_Alonso"
          target="_blank"
        >
          F. Alonso
        </a>
      </mdlistitem-stub>
      <mdlistitem-stub
        class=""
        data-index="12"
        dataindex="12"
      >
        
        13. 
        <strong>
          Hungarian Grand Prix
        </strong>
        
         - 
        
        <a
          href="http://en.wikipedia.org/wiki/Kimi_R%C3%A4ikk%C3%B6nen"
          target="_blank"
        >
          K. Räikkönen
        </a>
      </mdlistitem-stub>
      <mdlistitem-stub
        class=""
        data-index="13"
        dataindex="13"
      >
        
        14. 
        <strong>
          Turkish Grand Prix
        </strong>
        
         - 
        
        <a
          href="http://en.wikipedia.org/wiki/Kimi_R%C3%A4ikk%C3%B6nen"
          target="_blank"
        >
          K. Räikkönen
        </a>
      </mdlistitem-stub>
      <mdlistitem-stub
        class=""
        data-index="14"
        dataindex="14"
      >
        
        15. 
        <strong>
          Italian Grand Prix
        </strong>
        
         - 
        
        <a
          href="http://en.wikipedia.org/wiki/Juan_Pablo_Montoya"
          target="_blank"
        >
          J. Pablo Montoya
        </a>
      </mdlistitem-stub>
      <mdlistitem-stub
        class=""
        data-index="15"
        dataindex="15"
      >
        
        16. 
        <strong>
          Belgian Grand Prix
        </strong>
        
         - 
        
        <a
          href="http://en.wikipedia.org/wiki/Kimi_R%C3%A4ikk%C3%B6nen"
          target="_blank"
        >
          K. Räikkönen
        </a>
      </mdlistitem-stub>
      <mdlistitem-stub
        class=""
        data-index="16"
        dataindex="16"
      >
        
        17. 
        <strong>
          Brazilian Grand Prix
        </strong>
        
         - 
        
        <a
          href="http://en.wikipedia.org/wiki/Juan_Pablo_Montoya"
          target="_blank"
        >
          J. Pablo Montoya
        </a>
      </mdlistitem-stub>
      <mdlistitem-stub
        class=""
        data-index="17"
        dataindex="17"
      >
        
        18. 
        <strong>
          Japanese Grand Prix
        </strong>
        
         - 
        
        <a
          href="http://en.wikipedia.org/wiki/Kimi_R%C3%A4ikk%C3%B6nen"
          target="_blank"
        >
          K. Räikkönen
        </a>
      </mdlistitem-stub>
      <mdlistitem-stub
        class="highlight"
        data-index="18"
        dataindex="18"
      >
        
        19. 
        <strong>
          Chinese Grand Prix
        </strong>
        
         - 
        
        <a
          href="http://en.wikipedia.org/wiki/Fernando_Alonso"
          target="_blank"
        >
          F. Alonso
        </a>
      </mdlistitem-stub>
    </span>
  </mdlist-stub>
   
  <mddialog-stub
    mdbackdrop="true"
    mdbackdropclass="md-dialog-overlay"
    mdclickoutsidetoclose="true"
    mdcloseonesc="true"
    mdfullscreen="true"
  >
    <mddialogtitle-stub>
      Error
    </mddialogtitle-stub>
     
    <mddialogcontent-stub>
      
    </mddialogcontent-stub>
     
    <mddialogactions-stub>
      <mdbutton-stub
        type="button"
      >
        Ok
      </mdbutton-stub>
    </mddialogactions-stub>
  </mddialog-stub>
</div>
`;
